### PR TITLE
[retiarii] document

### PR DESCRIPTION
**Pull request by QuanluZhang**

### CODE REVIEW

File: docs/en_US/NAS/retiarii/Tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,171 @@
+Neural Architecture Search with Retiarii (Experimental)
+=======================================================
+
+`Retiarii <https://www.usenix.org/system/files/osdi20-zhang_quanlu.pdf>`__ is a new framework to support neural architecture search and hyper-parameter tuning. It allows users to express various search space with high flexibility, to reuse many SOTA search algorithms, and to leverage system level optimizations to speed up the search process.
+
```
Comment (colorjam) [inline docs/en_US/NAS/retiarii/Tutorial.rst:5]: The support of NAS and hyper-parameter tuning looks similar to NNI? Why highlight the new framework?

File: docs/en_US/NAS/retiarii/Tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+Define Model Mutations
+^^^^^^^^^^^^^^^^^^^^^^
+
+A base model is only one concrete model not a model space. To define model space, we provide APIs and primitives for users to express how the base model can be mutated.
+
+**Express mutations in an inlined manner**
+
+For easy usability and also backward compatibility, we provide some APIs for users to easily express possible mutations during defining a base model. The APIs can be used just like PyTorch module.
+
+* ``nn.LayerChoice``. It allows users to put several candidate operations (e.g., PyTorch modules), one is chosen in each explored model.
```
Comment (colorjam) [inline docs/en_US/NAS/retiarii/Tutorial.rst:61]: Which ``nn``  should be used? ``torch.nn`` or ``retiarii.nn.pytorch``?

Comment (QuanluZhang) [reply docs/en_US/NAS/retiarii/Tutorial.rst:61]: it should be `retiarii.nn.pytorch`, i will make it clear, thanks!

File: docs/en_US/NAS/retiarii/Tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+  applied_mutators = []
+  applied_mutators.append(BlockMutator('mutable_0'))
+  applied_mutators.append(BlockMutator('mutable_1'))
+
+``BlockMutator`` could be defined by users to express how to mutate the base model. User defined mutator should inherit ``Mutator`` class, and implement mutation logic in member function ``mutate``.
+
+.. code-block:: python
+
+  class BlockMutator(Mutator):
```
Comment (colorjam) [inline docs/en_US/NAS/retiarii/Tutorial.rst:103]: import Mutator?

File: docs/en_US/NAS/retiarii/Tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        super(BlockMutator, self).__init__()
+        self.target = target
+
+    def mutate(self, model):
+      nodes = model.get_nodes_by_label(self.target)
+      for node in nodes:
+        chosen_op = self.choice(candidate_op_list)
+        node.update_operation(chosen_op.type, chosen_op.params)
+
+The input of ``mutate`` is a model IR (please refer to `here <>`__ for the format and APIs of the IR), users can mutate the model with its member functions (e.g., ``get_nodes_by_label``, ``update_operation``). The mutation operations can be combined with the API ``self.choice``, in order to express a set of mutations. In the above example, the node's operation can be changed to each operation from ``candidate_op_list``.
```
Comment (colorjam) [inline docs/en_US/NAS/retiarii/Tutorial.rst:114]: What is model `IR`?

File: docs/en_US/NAS/retiarii/Tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+Users can write their own trainer and strategy, please refer to `here <>`__ for tutorial.
+
+Set up an Experiment
+^^^^^^^^^^^^^^^^^^^^
+
+After all the above are prepared, it is time to start an experiment to do the model search. We design unified interface for users to start their experiment. An example is shown below
+
+.. code-block:: python
+
+  exp = RetiariiExperiment(base_model, trainer, applied_mutators, simple_startegy)
```
Comment (colorjam) [inline docs/en_US/NAS/retiarii/Tutorial.rst:158]: import `RetiariiExperiment`?

File: docs/en_US/NAS/retiarii/Tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+  exp_config = RetiariiExeConfig('local')
```
Comment (colorjam) [inline docs/en_US/NAS/retiarii/Tutorial.rst:159]: import `RetiariiExperiment` and `RetiariiExeConfig`?

File: docs/en_US/NAS/retiarii/Tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+In this approach, trainer is for training each explored model, while strategy is for sampling the models. Both trainer and strategy are required to explore the model space.
+
+**Weight-sharing search approach:**
+In this approach, users only need a weight-sharing trainer, because this trainer takes charge of both search and training.
+
+In the following table, we listed the available trainers and strategies.
+
+TODO: table here.
+
+Users can write their own trainer and strategy, please refer to `here <>`__ for tutorial.
```
Comment (Lijiaoa) [inline docs/en_US/NAS/retiarii/Tutorial.rst:149]: null link.

File: docs/en_US/NAS/retiarii/Tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+**Classic search approach:**
+In this approach, trainer is for training each explored model, while strategy is for sampling the models. Both trainer and strategy are required to explore the model space.
+
+**Weight-sharing search approach:**
+In this approach, users only need a weight-sharing trainer, because this trainer takes charge of both search and training.
+
+In the following table, we listed the available trainers and strategies.
+
+TODO: table here.
```
Comment (Lijiaoa) [inline docs/en_US/NAS/retiarii/Tutorial.rst:147]: seems like that table should be here.

File: docs/en_US/NAS/retiarii/Tutorial.rst (hunk 4 above)
Comment (Lijiaoa) [inline docs/en_US/NAS/retiarii/Tutorial.rst:114]: also ` here ` is null

File: docs/en_US/NAS/retiarii/Tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+.. code-block:: python
+
+  # declared in `__init__`
+  self.input_switch = nn.InputChoice(n_chosen=1)
+  # invoked in `forward` function, choose one from the three
+  out = self.input_switch([tensor1, tensor2, tensor3])
+
+* ``nn.ValueChoice``. Will be supported soon.
+
+Detailed API description can be found `here <>`__\. Example of using these APIs can be found in `Darts base model <https://github.com/microsoft/nni/blob/master/test/retiarii_test/darts/darts_model.py>`__\.
```
Comment (Lijiaoa) [inline docs/en_US/NAS/retiarii/Tutorial.rst:85]: this link is null.

Comment (ultmaster) [inline docs/en_US/NAS/retiarii/Tutorial.rst:85]: Please use :githublink: directive

File: docs/en_US/NAS/retiarii/Tutorial.rst [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+  class BlockMutator(Mutator):
+    def __init__(self, target: str):
+        super(BlockMutator, self).__init__()
+        self.target = target
+
+    def mutate(self, model):
+      nodes = model.get_nodes_by_label(self.target)
+      for node in nodes:
+        chosen_op = self.choice(candidate_op_list)
```
Comment (ultmaster) [inline docs/en_US/NAS/retiarii/Tutorial.rst:111]: This is basically LayerChoice. I suggest we find a better example.

File: docs/en_US/NAS/retiarii/Tutorial.rst (hunk 1 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/retiarii/Tutorial.rst:5]: this is a good question. it provide new user experience, experiment is launched in python code, search space can be expressed directly in user model code, supports both neural architecture space and hyper-parameter space. will make it clear in this doc

File: docs/en_US/NAS/retiarii/Tutorial.rst (hunk 3 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/retiarii/Tutorial.rst:114]: imported, thanks

File: docs/en_US/NAS/retiarii/Tutorial.rst (hunk 4 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/retiarii/Tutorial.rst]: explained above

Comment (QuanluZhang) [reply docs/en_US/NAS/retiarii/Tutorial.rst]: fixed

File: docs/en_US/NAS/retiarii/Tutorial.rst (hunk 8 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/retiarii/Tutorial.rst]: added

File: docs/en_US/NAS/retiarii/Tutorial.rst (hunk 6 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/retiarii/Tutorial.rst:221]: no need to add, I just added the link to the complete code file.

**Review by Lijiaoa — APPROVED** [review]

**Comment by Lijiaoa** [comment]
Approve for null `here` link.

**Review by ultmaster — APPROVED** [review]

File: docs/en_US/NAS/retiarii/Tutorial.rst [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+
+Define Base Model
+^^^^^^^^^^^^^^^^^
+
+Defining a base model is almost the same as defining a PyTorch (or TensorFlow) model. There are only two small differences.
+
+* Use our wrapped ``nn`` for PyTorch modules instead of ``torch.nn``. Specifically, users can simply replace the code ``import torch.nn as nn`` with ``import nni.retiarii.nn.pytorch as nn``
+* Add the decorator ``@blackbox_module`` to some module classes. Below we explain why this decorator is needed and what module classes should be decorated.
+
+**@blackbox_module**: To understand this decorator, we first briefly explain how our framework works: it converts user defined model to a graph representation (called graph IR), each instantiated module is converted to a subgraph. Then user defined mutations are applied to the graph to generate new graphs. Each new graph is then converted back to PyTorch code and executed. ``@blackbox_module`` here means the module will not be converted to a subgraph but is converted to a single graph node. That is, the module will not be unfolded anymore. Users should/can decorate a module class in the following cases:
```
Comment (colorjam) [inline docs/en_US/NAS/retiarii/Tutorial.rst:29]: is this decorator an advanced usage? is it proper to mention it in this tutorial?

Comment (QuanluZhang) [reply docs/en_US/NAS/retiarii/Tutorial.rst:29]: it is not advanced usage...

**Review by liuzhe-lz — APPROVED** [review]

**Pull request merged**